src/backend/gporca/libgpopt/src/mdcache/CMDAccessorUtils.cpp

```
@@ -94,11 +94,12 @@ CMDAccessorUtils::FCmpExists(CMDAccessor *md_accessor, IMDId *left_mdid,
 	}
 	GPOS_CATCH_EX(ex)
 	{
-		GPOS_ASSERT(
-			GPOS_MATCH_EX(ex, gpdxl::ExmaMD, gpdxl::ExmiMDCacheEntryNotFound));
```

Previously, if the `GPOS_ASSERT` fails here, wouldn't it throw exception as well? In that case wouldn't the query be canceled as well?

---

Yes, if you had asserts enabled, then it should raise an exception, however it won't cancel the query.

See `CGPOptimizer::GPOPTOptimizedPlan()` exception handler for details.  A `GPOS_ASSERT` exception isn't handled, so ORCA will fail to create plan and fall back to PLANNER.  However, if it's a GPDB ERROR (e.g. cancel query ERROR), then handler will re-raise and eventually cancel the query.

---

I see. So all `GPOS_ASSERT` inside of a `GPOS_CATCH_EX` block needs to be handle this way, correct? And this PR covers all of them?

---

I think using `GPOS_ASSERT` should be nothing more than a developer tool to validate some assumptions regardless of the context. In which case, I think it's okay to use it inside of `GPOS_CATCH_EX`.

However, in this particular case I believe the assumption that we should only expect to see `ExmiMDCacheEntryNotFound` exceptions was incorrect.  If we haven't built with `GPOS_DEBUG` then we end up essentially swallowing any exception we didn't expect to see.  That seems incorrect.

I've updated all the `GPOS_CATCH_EX` blocks that made this assumption.

---

I see. That makes sense. With the fault injected, what would be the exception type here? `CException::ExmiOOM`?

---

The fault injected type is `GPOS_MATCH_EX(ex, gpdxl::ExmaGPDB, gpdxl::ExmiGPDBError)`; the same type as what the runaway cleaner ERROR is treated as.